cmd/snap-confine/ns-support.c

```diff
@@ -289,7 +289,14 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
 	}
 	for (mie = sc_first_mountinfo_entry(mi); mie != NULL;
 	     mie = sc_next_mountinfo_entry(mie)) {
-		if (!sc_streq(mie->mount_dir, "/")) {
+		// If the "/" directory is a tmpfs, then we'll look at the /usr
+		// directory
+		if (sc_streq(mie->fs_type, "tmpfs")) {
+			continue;
+		}
+
+		if (!sc_streq(mie->mount_dir, "/") &&
+			!sc_streq(mie->mount_dir, "/usr")) {
```

Comment:
is this check still correct if the snap layout adds a new directory under /usr? just modified test-snapd-layout to add a layout entry under `/usr/foo`, which then makes /usr be reported as tmpfs:
```
$ stat -f /usr
  File: "/usr"
    ID: 6585fe8765d1a97e Namelen: 255     Type: tmpfs
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 4100692    Free: 4100692    Available: 4100692
Inodes: Total: 4100692    Free: 4100681
```

As far as path prefix goes, the layout was not allowed to create top level directories only. Perhaps we could look at `/writable` as it makes no sense to modify that one with layouts?

Reply:
Thanks Maciek, you got a point there! I've prepared a change to deal with this (https://paste.ubuntu.com/p/rMtBdTfm5K/), but then while testing it (adding a layout for `/usr/foo`) I noticed that when the /usr tmpfs mimic is created, the original /usr device is still listed in mountinfo:
```
...
631 593 7:1 / /var/lib/snapd/hostfs/snap/core20/1518 ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
...
649 592 0:53 / / rw,relatime - tmpfs none rw,mode=770
650 649 7:1 /usr /usr ro,nodev,relatime master:396 - squashfs /dev/loop1 ro     <--- THIS LINE
...
617 650 0:54 / /usr rw,relatime - tmpfs tmpfs rw,mode=755
618 617 7:1 /usr/bin /usr/bin ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
619 617 7:1 /usr/games /usr/games ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
620 617 7:1 /usr/include /usr/include ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
...
615 617 7:2 /bin/cmd /usr/foo ro,nodev,relatime master:404 - squashfs /dev/loop2 ro
```

So the branch currently pushed here should handle the situation correctly. Though I'm not sure we want to rely on the existance of that `/usr` -> `/usr` bind mount, as it's a bit too subtle IMHO. So, if you agree, I'll push the change from the pastebin.

Reply:
The diff looks good. Please take a look at the formatting, I think `make fmt` should do the trick.